File: src/main/js/funk/Product.js
funk.Product = (function(){
    var makeString = function(product, separator){
        var total = product.productArity();

        var buffer = "";
        for(var i=0; i<total; i++) {
            var element = product.productElement(i);
            buffer += funk.has(element, 'toString') ? element.toString() : element;
            if(i < total - 1) {
                buffer += separator;
            }
        }
        return buffer;
    };

    var ProductImpl = function(){};
    ProductImpl.prototype = {};
    ProductImpl.prototype.constructor = ProductImpl;
    ProductImpl.prototype.productArity = function(){
        funk.util.isAbstract();
    };
    ProductImpl.prototype.productElement = function(index){
        funk.util.isAbstract();
    };
    ProductImpl.prototype.productPrefix = function(){
        return "";
    };
    ProductImpl.prototype.equals = function(value){
        if(value === funk.util.verifiedType(value, funk.Product)) {
            if(this.productArity() == value.productArity()) {
                var index = this.productArity();
                while(--index > -1) {
                    if(funk.util.ne(this.productElement(index), value.productElement(index))) {
                        return false;
                    }
                }
                return true;
            }
        }
        return false;
    };
    ProductImpl.prototype.getIterator = function(){
        return new funk.ProductIterator(this);
    };
    ProductImpl.prototype.toString = function(){
        if(0 == this.productArity()) {
            return this.productPrefix();
        }
        return this.productPrefix() + "(" + makeString(this, ", ") + ")";
    };
    ProductImpl.prototype.name = "Product";
    return ProductImpl;
})();

File: src/main/dart/funk/Product.dart
interface IProduct<T> extends Iterable<T>, IFunkObject default Product {
  
  T productElement(int index);

  int get productArity();

  String get productPrefix();
}


class Product<T> implements IProduct<T> {
  
  Product(){
  }
  
  abstract T productElement(int index);
  
  String _makeString(String separator) {
    final int total = productArity;
    
    StringBuffer buffer = new StringBuffer();
    
    for(int i = 0; i < total; i++){
      final T element = productElement(i);
      buffer.add(element.toString());
      if(i < total - 1) {
        buffer.add(separator);
      }
    }
    
    return buffer.toString();
  }
  
  int hashCode() {
    int result = 17;
    int n = productArity;
    for(int i = 0; i < n; i++) {
      var item = productElement(i);
      if(item is Hashable) {
        result += item.hashCode();
      }
    }
    return result;
  }
  
  bool equals(IFunkObject that) {
    if(this == that) {
     return true; 
    }
    
    return hashCode() == that.hashCode();
  }
  
  abstract int get productArity();
  
  abstract String get productPrefix();
  
  Iterator<T> iterator() => new ProductIterator<T>(this);
  
  String toString(){
    if(0 == productArity) {
      return productPrefix;
    }
    return "${productPrefix}(${_makeString(', ')})";
  }
}
